src/server/query.c

```diff
@@ -65,6 +66,58 @@ tds_lastpacket(TDSSOCKET * tds)
 	return tds->in_buf[1] != 0;
 }
 
+/**
+ * get query packet of a given type
+ * \tds
+ * \param head         extra information to put in a TDS7 header
+ */
+static TDSRET
+tds_get_query_head(TDSSOCKET * tds, TDSHEADERS * head)
+{
+	if (!IS_TDS72_PLUS(tds->conn)) {
+		return TDS_SUCCESS;
+	}
+
+	int qn_len = 0, more;
+	char *qn_msgtext = NULL;
+	char *qn_options = NULL;
+	size_t qn_msgtext_len = 0;
+	size_t qn_options_len = 0;
+
+	qn_len = tds_get_int(tds) - 4 - 18;             /* total length */
+	tds_get_int(tds);                          /* length: transaction descriptor, ignored */
+	tds_get_smallint(tds);                      /* type: transaction descriptor, ignored */
+	tds_get_n(tds, tds->conn->tds72_transaction, 8);  /* transaction */
+	tds_get_int(tds);                           /* request count, ignored */
+	if (qn_len != 0) {
+		if (!head) {
+			tds_set_state(tds, TDS_IDLE);
+			return TDS_FAIL;
+		}
+
+		qn_len = tds_get_int(tds);                   /* length: query notification */
+		tds_get_smallint(tds);                      /* type: query notification, ignored */
+		qn_msgtext_len = tds_get_smallint(tds);  /* notifyid */
+		if (qn_msgtext_len > 0) {
+			qn_msgtext = (char *) realloc(qn_msgtext, qn_msgtext_len);
+			tds_get_n(tds, qn_msgtext, qn_msgtext_len);
+		}
+
+		qn_options_len = tds_get_smallint(tds);  /* ssbdeployment */
+		if (qn_options_len > 0) {
+			qn_options = (char *) realloc(qn_options, qn_options_len);
```

Comment:
This code will crash if `realloc` returns NULL. Other part is handling these conditions. Maybe for the server just using a kind of `xrealloc` would be enough.

Reply:
was mimic code from tds/query.c
tds_start_query_head(TDSSOCKET *tds, unsigned char packet_type, TDSHEADERS * head)

where was just using realloc. all realloc would have the same issue.

Reply:
It's a bit difference between checking for NULL handling that case returning error and ignoring and use as it were no NULL causing a crash.

Reply:
I switch all out with xrealloc. I wanted to try and build some unittests.

Also noticed a few things that would need to be corrected. But probably better incrementally.

Thank you for working with me freddy77. better then being in a vacuum.